README
parprouted, (C) 2002 Vladimir Ivaschenko <vi@maks.net>
-------------------------------------------------------

1) About:

parprouted is a daemon for transparent IP (Layer 3) proxy ARP bridging. 
Unlike standard bridging, proxy ARP bridging allows to bridge Ethernet 
networks behind wireless nodes. Normal L2 bridging does not work between 
wireless nodes because wireless does not know about MAC addresses used 
in the wired Ethernet networks.

The daemon will check ARP table every second and add direct /32 routes to your
kernel to the hosts learned via ARP. This way, your kernel will be able
to route the packets to the destination host when it receives it and you will
not have to care about setting routing/subnetting manually. All routes
entered by the daemon will have a metric of 50. 

2) Installation:

parprouted requires "ip" program form iproute2 tools to be installed in 
/sbin. If it is installed in another location, please replace "/sbin/ip" 
occurances in the source with the correct path.

In order for this daemon to work, you have to enable proxy ARP in your 
kernel. You can do it like this:

cat 1 > /proc/sys/net/conf/[interface]/proxy_arp

You have to do it for each interface you want to bridging on.

The daemon accept only one switch "-d", which stands for debugging. If you
run it in debugging mode the daemon will go to background and will print
additional debugging information to stdout/stderr.


CHANGELOG
Sun 11 Jan 2004 01:38:43 PM EET
parprouted 0.6:
Now ARP queries from the local host are resolved as well. Various small updates
to code and documentation.

------------------------------------------------------------------------
Tue 06 Jan 2004 09:22:09 PM EET
parprouted 0.50:
Much improved performance and robustness. Request queue is now kept
for all received ARP requests, ensuring that all new requests get answered
quickly (around 60 ms in my test).

------------------------------------------------------------------------
Wed 18 Sep 2002 11:10:41 AM EEST
parprouted 0.42:
Fixed incorrect route installation using unfinished ARP entry.
More aggressive use of mutexes.

------------------------------------------------------------------------
Sun 07 Jul 2002 10:20:16 PM EEST

parprouted 0.4:
Refresh ARP entries by ARP pinging them. Or make permanent ARP
entries if "-p" switch is given.

------------------------------------------------------------------------
Sun 07 Jul 2002 12:51:36 AM EEST

parprouted 0.3:
Critical bugfix. Previous parprouted failed in networks of any 
considerable complexity. Almost complete rewrite. parprouted now is 
multithreaded and does  everything in userspace (including ARP relaying,
spoofing and listening). No need to set proxy_arp in kernel anymore. 

arp.c
/* Code partly based on tarpd.c 1.6 (Public Domain):
 * 	$Id: tarpd.c,v 1.6 1999/09/17 12:09:40 tricky Exp $
 */

/* (C) 2002 Vladimir Ivaschenko <vi@maks.net>
 *
 * This application is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <net/if.h>
#include <net/ethernet.h>
#include <net/if_arp.h>
#include <linux/if_packet.h>
#include <netinet/if_ether.h>
#include <sys/ioctl.h>

#include "parprouted.h"

typedef struct _ether_arp_frame { 
  struct ether_header ether_hdr;
  struct ether_arp arp;
} ether_arp_frame;

int ipaddr_known(ARPTAB_ENTRY *list, struct in_addr addr, char *ifname) 
{

  while (list != NULL) {
    /* If we have this address in the table and ARP request comes from a 
       different interface, then we can reply */
    if ( addr.s_addr == list->ipaddr_ia.s_addr && strcmp(ifname, list->ifname)) {
	pthread_mutex_unlock(&arptab_mutex);
        return 1;
    }
    list = list->next;
  }
  
  if (debug)
      printf ("Did not find match for %s(%s)\n", inet_ntoa(addr), ifname);

  return 0;
}

int arp_recv(int sock, ether_arp_frame *frame) 
{
    char packet[4096];
    int nread;

    memset(frame, 0, sizeof(ether_arp_frame));
    
    nread=recv(sock, &packet, sizeof(packet), 0);
    
    if (nread > sizeof(ether_arp_frame)) {
	nread=sizeof(ether_arp_frame);
    }
    
    memcpy(frame, &packet, nread); 

    return nread;
}

/* Send ARP is-at reply */

void arp_reply(int sock, ether_arp_frame *frame, struct sockaddr_ll *ifs) 
{
  struct ether_arp *arp = &frame->arp;
  unsigned char ip[4];

  memcpy(&frame->ether_hdr.ether_dhost, &arp->arp_sha, ETH_ALEN);
  memcpy(&frame->ether_hdr.ether_shost, ifs->sll_addr, ETH_ALEN);

  memcpy(&arp->arp_tha, &arp->arp_sha, ETH_ALEN);
  memcpy(&arp->arp_sha, ifs->sll_addr, ETH_ALEN);

  memcpy(ip, &arp->arp_spa, 4);
  memcpy(&arp->arp_spa, &arp->arp_tpa, 4);
  memcpy(&arp->arp_tpa, ip, 4);

  arp->arp_op = htons(ARPOP_REPLY);

  sendto(sock, frame, sizeof(ether_arp_frame), 0, 
	 (struct sockaddr *)ifs, sizeof(struct sockaddr_ll));
}

/* Send ARP who-has request */

void arp_req(char *ifname, struct in_addr remaddr)
{
  ether_arp_frame frame;
  struct ether_arp *arp = &frame.arp;
  int sock;
  struct sockaddr_ll ifs;
  struct ifreq ifr;
  unsigned long ifaddr; 
  struct sockaddr_in *sin;
  
  sock = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ARP));

  /* Get the hwaddr and ifindex of the interface */
  memset(ifr.ifr_name, 0, IFNAMSIZ);
  strncpy(ifr.ifr_name, (char *) ifname, IFNAMSIZ);
  if(ioctl(sock, SIOCGIFHWADDR, &ifr) < 0) {
    syslog(LOG_ERR, "error: ioctl SIOCGIFHWADDR for %s: %s\n", (char *) ifname, sys_errlist[errno]);
    abort();
  }

  memset(ifs.sll_addr, 0, ETH_ALEN);
  memcpy(ifs.sll_addr, ifr.ifr_hwaddr.sa_data, ETH_ALEN);

  if(ioctl(sock, SIOCGIFINDEX, &ifr) < 0) {
	syslog(LOG_ERR, "error: ioctl SIOCGIFINDEX for %s: %s", (char *) ifname, sys_errlist[errno]);
        return;
  }

  ifs.sll_family = AF_PACKET;
  ifs.sll_protocol = htons(ETH_P_ARP);
  ifs.sll_ifindex = ifr.ifr_ifindex;
  ifs.sll_hatype = ARPHRD_ETHER;
  ifs.sll_pkttype = PACKET_BROADCAST;
  ifs.sll_halen = ETH_ALEN;

  if (ioctl(sock, SIOCGIFADDR, &ifr) == 0) {
	sin = (struct sockaddr_in *) &ifr.ifr_addr;
	ifaddr = sin->sin_addr.s_addr;
  } else {
	syslog(LOG_ERR, "error: ioctl SIOCGIFADDR for %s: %s", (char *) ifname, sys_errlist[errno]);
	return;
  }

  memset(&frame.ether_hdr.ether_dhost, 0xFF, ETH_ALEN);
  memcpy(&frame.ether_hdr.ether_shost, ifs.sll_addr, ETH_ALEN);
  frame.ether_hdr.ether_type = htons(ETHERTYPE_ARP);

  arp->arp_hrd = htons(ARPHRD_ETHER);
  arp->arp_pro = htons(ETH_P_IP);
  arp->arp_hln = 6;
  arp->arp_pln = 4;
  memset(&arp->arp_tha, 0, ETH_ALEN);
  memcpy(&arp->arp_sha, ifs.sll_addr, ETH_ALEN);

  memcpy(&arp->arp_tpa, &remaddr.s_addr, 4);
  memcpy(&arp->arp_spa, &ifaddr, 4);

  arp->arp_op = htons(ARPOP_REQUEST);

  if (debug) 
      printf("Relaying ARP request for %s to %s\n", inet_ntoa(remaddr), ifname);
  sendto(sock, &frame, sizeof(ether_arp_frame), 0, 
	 (struct sockaddr *) &ifs, sizeof(struct sockaddr_ll));
  close(sock);
}

void send_dummy_udp(char *ifname, u_int32_t remaddr) {
    int sock;
    struct sockaddr_in target;
    char sndbuf;
    
    sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
	syslog(LOG_ERR, "send_dummy_udp: socket error: %s", sys_errlist[errno]);
	return;
    }

    if (setsockopt(sock, SOL_SOCKET, SO_BINDTODEVICE, ifname, strlen(ifname)+1) == -1) {
	syslog(LOG_ERR, "send_dummy_udp: couldn't bind to device %s: %s", ifname, sys_errlist[errno]);
	return;
    }
        
    target.sin_family = AF_INET;
    memcpy(&target.sin_addr, &remaddr, sizeof(remaddr));
    target.sin_port = htons(44444);
    sendto(sock, &sndbuf, sizeof(sndbuf), 0, (struct sockaddr *) &target, sizeof(target));

    close(sock);
}

/* ARP ping all entries in the table */

void refresharp(ARPTAB_ENTRY *list)
{
  pthread_mutex_lock(&arptab_mutex);

  if (debug) 
      printf("Refreshing ARP entries.\n");
      
  while(list != NULL) {
    arp_req(list->ifname, list->ipaddr_ia);
    list = list->next;
  }

  pthread_mutex_unlock(&arptab_mutex);
}

void *arp(void *ifname) 
{
  int sock,i;
  struct sockaddr_ll ifs;
  struct ifreq ifr;

  pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
  pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL);

  sock = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ARP));

  if(sock == -1) {
    fprintf(stderr, "Socket error %d.\n", errno);
    exit(1);
  }

  /* Get the hwaddr and ifindex of the interface */
  memset(ifr.ifr_name, 0, IFNAMSIZ);
  strncpy(ifr.ifr_name, (char *) ifname, IFNAMSIZ);
  if(ioctl(sock, SIOCGIFHWADDR, &ifr) < 0) {
    syslog(LOG_ERR, "error: ioctl SIOCGIFHWADDR for %s: %s\n", (char *) ifname, sys_errlist[errno]);
    abort();
  }

  memset(ifs.sll_addr, 0, ETH_ALEN);
  memcpy(ifs.sll_addr, ifr.ifr_hwaddr.sa_data, ETH_ALEN);

  if(ioctl(sock, SIOCGIFINDEX, &ifr) < 0) {
    syslog(LOG_ERR, "error: ioctl SIOCGIFINDEX for %s: %s", (char *) ifname, sys_errlist[errno]);
    abort();
  }

  ifs.sll_family = AF_PACKET;
  ifs.sll_protocol = htons(ETH_P_ARP);
  ifs.sll_ifindex = ifr.ifr_ifindex;
  ifs.sll_hatype = ARPHRD_ETHER;
  ifs.sll_pkttype = PACKET_BROADCAST;
  ifs.sll_halen = ETH_ALEN;
  
  if(bind(sock, (struct sockaddr *)&ifs, sizeof(struct sockaddr_ll)) < 0) {
    fprintf(stderr, "Bind %s: %d\n", (char *) ifname, errno);
    abort();
  }

  while (1) {
    ether_arp_frame frame;
    unsigned long src;
    unsigned long dst;
    struct in_addr sia;
    struct in_addr dia;

    do {
      pthread_testcancel();
      /* I just want to sleep a bit */
      usleep(300);

      if (arp_recv(sock, &frame) <= 0)
    	    continue;
      /* Insert all the replies into ARP table */
      if (frame.arp.arp_op == ntohs(ARPOP_REPLY)) {
	  struct arpreq k_arpreq;
	  int arpsock;
	  struct sockaddr_in *sin;
	  
	  if ((arpsock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
		syslog(LOG_ERR, "error: ARP socket for %s: %s", (char *) ifname, sys_errlist[errno]);
	        continue;
	  }
	  
          k_arpreq.arp_ha.sa_family = ARPHRD_ETHER;
          memcpy(&k_arpreq.arp_ha.sa_data, &frame.arp.arp_sha, sizeof(frame.arp.arp_sha));
	  k_arpreq.arp_flags = ATF_COM;
	  if (option_arpperm)
	      k_arpreq.arp_flags = k_arpreq.arp_flags | ATF_PERM;
	  strncpy(k_arpreq.arp_dev, ifname, sizeof(k_arpreq.arp_dev));

	  k_arpreq.arp_pa.sa_family = AF_INET;
	  sin = (struct sockaddr_in *) &k_arpreq.arp_pa;
	  memcpy(&sin->sin_addr.s_addr, &frame.arp.arp_spa, sizeof(sin->sin_addr));
	  
	  if (debug)
	      printf("Updating kernel ARP table for %s(%s).\n", inet_ntoa(sin->sin_addr), (char *) ifname);
	  if (ioctl(arpsock, SIOCSARP, &k_arpreq) < 0) {
		syslog(LOG_ERR, "error: ioctl SIOCSARP for %s(%s): %s", inet_ntoa(sin->sin_addr), (char *) ifname, sys_errlist[errno]);
		close(arpsock);
	        continue;
	  }
	   
	  close(arpsock);
      }
    } while (frame.arp.arp_op != htons(ARPOP_REQUEST));
    
    src = *((long *)frame.arp.arp_spa);
    dst = *((long *)frame.arp.arp_tpa);
    
    dia.s_addr = dst;
    sia.s_addr = src;
    
    if (debug)
	  printf("Received ARP request for %s on iface %s\n", inet_ntoa(dia), (char *) ifname);
    /* Relay ARP request to all other interfaces */
    for (i=0; i <= last_iface_idx; i++) {
	if (strcmp(ifaces[i], ifname)) {
	    arp_req(ifaces[i], dia);
            /* send_dummy_udp(ifaces[i], dst); */
	}
    }

    pthread_mutex_lock(&arptab_mutex);
    if ( ipaddr_known(*arptab, dia, (char *) ifname) != 0 ) {
      if (debug) {
          printf("Replying to %s faking ", inet_ntoa(sia));
	  printf("%s\n", inet_ntoa(dia));
      }
      arp_reply(sock, &frame, &ifs);
    }
    pthread_mutex_unlock(&arptab_mutex);
  }
}



parprouted.c
/* parprouted: ProxyARP routing daemon. 
 * (C) 2002 Vladimir Ivaschenko <vi@maks.net>
 *
 * This application is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */
 
#include "parprouted.h"

char *progname;
int debug=0;
char *errstr;

pthread_t my_threads[MAX_IFACES+1];
int last_thread_idx=-1;

char * ifaces[MAX_IFACES+2];
int last_iface_idx=-1;

ARPTAB_ENTRY **arptab;
pthread_mutex_t arptab_mutex;

ARPTAB_ENTRY * findentry(char *ipaddr) 
{
    ARPTAB_ENTRY * cur_entry=*arptab;
    ARPTAB_ENTRY * prev_entry=NULL;
        
    while (cur_entry != NULL && strcmp(ipaddr, cur_entry->ipaddr)) {
	prev_entry = cur_entry;
	cur_entry = cur_entry->next;
    };

    if (cur_entry == NULL) {
	if ((cur_entry = (ARPTAB_ENTRY *) malloc(sizeof(ARPTAB_ENTRY))) == NULL) {
	    errstr = strerror(errno);
	    syslog(LOG_INFO, "No memory: %s", errstr);
	} else {
	    if (prev_entry == NULL) { *arptab=cur_entry; }
	    else { prev_entry->next = cur_entry; }
	    cur_entry->next = NULL;
	    cur_entry->ipaddr[0] = '\0';
	    cur_entry->ifname[0] = '\0';
	    cur_entry->route_added=0;
	}
    }
    
    return cur_entry;	
}

void processarp(int cleanup) 
{
    ARPTAB_ENTRY *cur_entry=*arptab, *prev_entry=NULL;
    char routecmd_str[ROUTE_CMD_LEN];

    while (cur_entry != NULL) {
	if (cur_entry->tstamp-time(NULL) <= ARP_TABLE_ENTRY_TIMEOUT 
	    && !cur_entry->route_added && !cleanup) 
	{
	    /* added route to the kernel */
	    if (snprintf(routecmd_str, ROUTE_CMD_LEN-1, 
		     "/sbin/ip route add %s/32 metric 50 dev %s scope link",
		     cur_entry->ipaddr, cur_entry->ifname) > ROUTE_CMD_LEN-1) 
	    {
		syslog(LOG_INFO, "ip route command too large to fit in buffer!");
	    } else {
		if (system(routecmd_str) != 0)
		    { syslog(LOG_INFO, "'%s' unsuccessful!", routecmd_str); }
	    }

	    cur_entry->route_added = 1;
	    cur_entry = cur_entry->next;

	} else if (cur_entry->tstamp-time(NULL) > ARP_TABLE_ENTRY_TIMEOUT || cleanup) {
	    /* remove entry from arp table and remove route from kernel */
	    if (snprintf(routecmd_str, ROUTE_CMD_LEN-1, 
		     "/sbin/ip route del %s/32 metric 50 dev %s scope link",
		     cur_entry->ipaddr, cur_entry->ifname) > ROUTE_CMD_LEN-1) 
	    {
		syslog(LOG_INFO, "ip route command too large to fit in buffer!");
	    } else {
		if (system(routecmd_str) != 0)
		    syslog(LOG_INFO, "'%s' unsuccessful!", routecmd_str);
	    }

	    if (prev_entry != NULL) {
		prev_entry->next = cur_entry->next;
		free(cur_entry);
		cur_entry=prev_entry->next;
	    } else {
		*arptab = cur_entry->next;
		free(cur_entry);
		cur_entry=*arptab;
	    }
		
	} else {
	    cur_entry = cur_entry->next;
	} /* if */

    } /* while loop */
}	

void parseproc()
{
    FILE *arpf;
    int firstline;
    ARPTAB_ENTRY *entry;
    char line[ARP_LINE_LEN];
    char *item;

    /* Parse /proc/net/arp table */
        
    if ((arpf = fopen(PROC_ARP, "r")) == NULL) {
	errstr = strerror(errno);
	syslog(LOG_INFO, "Error during ARP table open: %s", errstr);
    }

    firstline=1;
    
    while (!feof(arpf)) {
	
	if (fgets(line, ARP_LINE_LEN, arpf) == NULL) {
	    if (!ferror(arpf))
		break;
	    else {
    		errstr = strerror(errno);
		syslog(LOG_INFO, "Error during ARP table open: %s", errstr);
	    }
	} else {
	    if (firstline) { firstline=0; continue; }
	    
	    item=strtok(line, " ");

	    pthread_mutex_lock(&arptab_mutex);
	    entry=findentry(item);
	    
	    if (strlen(item) < ARP_TABLE_ENTRY_LEN)
		strncpy(entry->ipaddr, item, ARP_TABLE_ENTRY_LEN);
	    else {
    		    errstr = strerror(errno);
		    syslog(LOG_INFO, "Error during ARP table parsing: %s", errstr);
	    }
	    if ((entry->ipaddr_ia.s_addr = inet_addr(entry->ipaddr)) == -1)
		    syslog(LOG_INFO, "Error parsing IP address %s", entry->ipaddr);
	    
	    item=strtok(NULL, " "); item=strtok(NULL, " "); item=strtok(NULL, " ");
	    if (strlen(item) < ARP_TABLE_ENTRY_LEN)
		strncpy(entry->hwaddr, item, ARP_TABLE_ENTRY_LEN);
	    else {
    		    errstr = strerror(errno);
		    syslog(LOG_INFO, "Error during ARP table parsing: %s", errstr);
	    }

	    item=strtok(NULL, " "); item=strtok(NULL, " ");
	    if (item[strlen(item)-1] == '\n') { item[strlen(item)-1] = '\0'; }
	    if (strlen(item) < ARP_TABLE_ENTRY_LEN) {
		if (entry->route_added && !strcmp(item, entry->ifname))
		    /* Remove route from kernel if it already exists */
		    entry->tstamp=0;
		else 
		    strncpy(entry->ifname, item, ARP_TABLE_ENTRY_LEN);
	    } else {
    		    errstr = strerror(errno);
		    syslog(LOG_INFO, "Error during ARP table parsing: %s", errstr);
	    }

	    time(&entry->tstamp);
	    
	    if (debug &! entry->route_added) {
	        printf("refresh entry: IPAddr: '%s' HWAddr: '%s' Dev: '%s'\n", 
		    entry->ipaddr, entry->hwaddr, entry->ifname);
	    }

	    pthread_mutex_unlock(&arptab_mutex);
	}
    }

    if (fclose(arpf)) {
	errstr = strerror(errno);
	syslog(LOG_INFO, "Error during ARP table open: %s", errstr);
    }
}

void cleanup() 
{
    int i=0;
    
    syslog(LOG_INFO, "Received signal; cleaning up.");
    for (i=0; i <= last_thread_idx; i++) {
    }
    pthread_mutex_trylock(&arptab_mutex);
    processarp(1);
    syslog(LOG_INFO, "Terminating.");
    exit(1);
}

void sighandler()
{
    pthread_exit(NULL);
}

void *main_thread()
{
    signal(SIGINT, sighandler);
    signal(SIGTERM, sighandler);
    signal(SIGHUP, sighandler);
    
    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
    pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL);
    pthread_cleanup_push(cleanup, NULL);
    while (1) {
	pthread_testcancel();
        parseproc();
        processarp(0);
	sleep(SLEEPTIME);
    }
    pthread_cleanup_pop(1);
}
    
int main (int argc, char **argv)
{
    pid_t child_pid;
    int i, help=1;
    
    progname = (char *) basename(argv[0]);
    
    for (i = 1; i < argc; i++) {
	if (!strcmp(argv[i],"-d")) { 
	    debug=1;
	    help=0;
	}
	else if (!strcmp(argv[i],"-h") || !strcmp(argv[i],"--help")) {
	    break;
	}
	else {
	    last_iface_idx++;
	    ifaces[last_iface_idx]=argv[i];
	    help=0;
	}
    }

    if (help || last_iface_idx <= -1) {
	    printf("parprouted: proxy ARP routing daemon, version %s.\n", VERSION);
    	    printf("(C) 2002 Vladimir Ivaschenko <vi@maks.net>, GPL2 license.\n");
	    printf("Usage: parprouted [-d] interface [interface]\n");
	    exit(1);
    }

    if (!debug) {
	/* stolen from watchdog.c */
        /* fork to go into the background */
        if ((child_pid = fork()) < 0) {
            perror(progname);
            exit(1);
        } else if (child_pid > 0) {
            /* fork was okay          */
            /* wait for child to exit */
            if (waitpid(child_pid, NULL, 0) != child_pid) {
                perror(progname);
                exit(1);
            }
            /* and exit myself */
            exit(0);
        }
        /* and fork again to make sure we inherit all rights from init */
        if ((child_pid = fork()) < 0) {
            perror(progname);
            exit(1);
        } else if (child_pid > 0)
            exit(0);

        /* Okay, we're a daemon     */
        /* but we're still attached to the tty */
        /* create our own session */
        setsid();

        /* with USE_SYSLOG we don't do any console IO */
        close(0);
        close(1);
        close(2);

    }

      
    openlog(progname, LOG_PID | LOG_CONS | LOG_PERROR, LOG_DAEMON);
    syslog(LOG_INFO, "Starting.");

    if ((arptab = (ARPTAB_ENTRY **) malloc(sizeof(ARPTAB_ENTRY **))) == NULL) {
	    errstr = strerror(errno);
	    syslog(LOG_INFO, "No memory: %s", errstr);
    }
    
    *arptab = NULL;

    if (pthread_create(&my_threads[++last_thread_idx], NULL, main_thread, NULL)) {
	syslog(LOG_ERR, "Error creating main thread.");
	abort();
    }

    signal(SIGINT, SIG_IGN);
    signal(SIGTERM, SIG_IGN);
    signal(SIGHUP, SIG_IGN);

    for (i=0; i <= last_iface_idx; i++) {
	if (pthread_create(&my_threads[++last_thread_idx], NULL, arp, (void *) ifaces[i])) {
	    syslog(LOG_ERR, "Error creating ARP thread for %s.",ifaces[i]);
	    abort();
	}
	if (debug) printf("Created ARP thread for %s.\n",ifaces[i]);
    }
        
    if (pthread_join(my_threads[0], NULL)) {
	syslog(LOG_ERR, "Error joining thread.");
	abort();
    }

    while (waitpid(-1, NULL, WNOHANG)) { }    
    exit(1);     
}


parprouted.h
/* parprouted: ProxyARP routing daemon. 
 * (C) 2002 Vladimir Ivaschenko <vi@maks.net>
 *
 * This application is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#define PROC_ARP "/proc/net/arp"
#define ARP_LINE_LEN 255
#define ARP_TABLE_ENTRY_LEN 20
#define ARP_TABLE_ENTRY_TIMEOUT 60 /* seconds */
#define ROUTE_CMD_LEN 255
#define SLEEPTIME 1000000 /* ms */
#define REFRESHTIME 50 /* seconds */
#define MAX_IFACES 10

#define MAX_RQ_SIZE 50	/* maximum size of request queue */

#define VERSION "0.6"

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>
#include <sys/wait.h>
#include <string.h>
#include <time.h>
#include <stdlib.h>
#include <assert.h>
#include <libgen.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>
#include <stdlib.h>
#include <unistd.h>

typedef struct arptab_entry {
    struct in_addr ipaddr_ia;
    char hwaddr[ARP_TABLE_ENTRY_LEN];
    char ifname[ARP_TABLE_ENTRY_LEN];
    time_t tstamp;
    int route_added;
    int incomplete;
    struct arptab_entry *next;
} ARPTAB_ENTRY;

extern int debug;
extern int option_arpperm;

extern ARPTAB_ENTRY **arptab;
extern pthread_mutex_t arptab_mutex;
extern pthread_mutex_t req_queue_mutex;

extern char * ifaces[MAX_IFACES+2];
extern int last_iface_idx;

extern void *arp(char *ifname);
extern void refresharp(ARPTAB_ENTRY *list);
extern void arp_req(char *ifname, struct in_addr remaddr);

extern void parseproc();
extern void processarp(int cleanup);


